noir-projects/aztec-nr/aztec/src/keys/getters.nr

```diff
@@ -105,3 +111,63 @@ fn fetch_and_constrain_keys(address: AztecAddress) -> PublicKeys {
 pub fn get_nsk_app(npk_m_hash: Field) -> Field {
     get_key_validation_request(npk_m_hash, NULLIFIER_INDEX).sk_app
 }
+
+// This returns all current public keys for a given account, applying proper constraints to the context. We read all
+// keys at once since the contraints for reading them all are actually fewer than if we read them one at a time - any
+// read keys that are not required by the caller can simply be discarded.
+pub fn get_current_public_keys(context: &mut PrivateContext, account: AztecAddress) -> PublicKeys {
+    // The keys are stored in the key registry's public storage, but accessing that directly would require performing
+    // multiple historical reads, which results in a large number of constraints (since each field would result in a
+    // full merkle inclusion proof). Instead, we rely on the fact that the registry also stores the hash of all keys,
+    // and then show that this hash matches public keys we get from an unconstrained function.
+
+    let hinted_public_keys = key_registry_get_stored_keys_hint(account);
+
+    let base_storage_slot = derive_storage_slot_in_map(KEY_REGISTRY_STORAGE_SLOT, account);
+    // The keys are stored as [ ...serialized_keys, hash ], and since arrays get allocated sequential storage slots 
+    // (prior to siloing!), we simply add the length to the base slot to get the last element.
+    let hash_storage_slot = base_storage_slot + PUBLIC_KEYS_LENGTH as Field;
+
+    // Here we prove that the unconstrained keys are the correct preimage of the hash stored in public storage, and
+    // hence they are the correct keys.
+    let historical_header = context.get_header();
+    // Note that the hash of zeroed keys is special-cased as also being zero, so we don't need to worry about that edge
+    // case: uninitialized key registry storage is valid, since it'll hold zeroed keys and the zero hash.
+    assert_eq(
+        hinted_public_keys.hash().inner, historical_header.public_storage_historical_read(hash_storage_slot, KEY_REGISTRY_CONTRACT_ADDRESS)
+    );
+
+    // We now need to constrain the caller so that they cannot use very old blocks when constructing proofs, and hence
+    // read very old keys. We are lax and allow _any_ recent block number to be used, regardless of whether there may
+    // have been a recent key rotation. This means that multiple sets of keys are valid for a while immediately after
+    // rotation, until the old keys become phased out. We *must* be lax to prevent denial of service and transaction
+    // fingerprinting attacks by accounts that rotate their keys frequently.
```

Comment:
I am not aware how exactly the fingerprinting attacks can be performed. Could you briefly describe it here?

Reply:
Imagine some accounts needs to be sent notes (e.g. they're a fee collector), and they rotate their keys frequently, choosing specific blocks of change. Transactions that intereact with whatever protocol they're collecting fees in might be identifiable by having that block of change.

Wallets could protect themselves regardless by setting the max block number by other means, but that's a different conversation.